akka-javasdk/src/main/scala/akka/javasdk/impl/workflow/ReflectiveWorkflowRouter.scala

```diff
@@ -4,63 +4,199 @@
-      val methodInvoker = commandHandler.getSingleNameInvoker()
-      val deserializedCommand =
-        CommandSerialization.deserializeComponentClientCommand(methodInvoker.method, scalaPbAnyCommand)
-      val result = deserializedCommand match {
-        case None          => methodInvoker.invoke(workflow)
-        case Some(command) => methodInvoker.invokeDirectly(workflow, command)
+      command: BytesPayload,
+      context: CommandContext,
+      timerScheduler: TimerScheduler): CommandResult = {
+
+    val commandEffect =
+      try {
+        workflow._internalSetup(decodeUserState(userState), context, timerScheduler)
+
+        val commandHandler = commandHandlerLookup(commandName)
+
+        // Commands can be in three shapes:
+        // - BytesPayload.empty - there is no real command, and we are calling a method with arity 0
+        // - BytesPayload with json - we deserialize it and call the method
+        // - BytesPayload with Proto encoding - we deserialize using InvocationContext
+        if (serializer.isJson(command) || command != BytesPayload.empty) {
+          // special cased component client calls, lets json commands through all the way
+          val methodInvoker = commandHandler.getSingleNameInvoker()
+          val deserializedCommand =
+            CommandSerialization.deserializeComponentClientCommand(methodInvoker.method, command, serializer)
+          val result = deserializedCommand match {
+            case None        => methodInvoker.invoke(workflow)
+            case Some(input) => methodInvoker.invokeDirectly(workflow, input)
+          }
+          result.asInstanceOf[Workflow.Effect[_]]
+        } else {
+
+          // FIXME can be proto from http-grpc-handling of the static es endpoints
+          val pbAnyCommand = AnySupport.toScalaPbAny(command)
+          val invocationContext =
+            InvocationContext(pbAnyCommand, commandHandler.requestMessageDescriptor, context.metadata())
+
+          val inputTypeUrl = pbAnyCommand.typeUrl
+
+          val methodInvoker = commandHandler.getInvoker(inputTypeUrl)
+          methodInvoker
+            .invoke(workflow, invocationContext)
+            .asInstanceOf[Workflow.Effect[_]]
+        }
+
+      } catch {
+        case CommandHandlerNotFound(name) =>
+          throw new WorkflowException(
+            context.workflowId(),
+            commandName,
+            s"No command handler found for command [$name] on ${workflow.getClass}")
+      } finally {
+        workflow._internalClear();
       }
-      result.asInstanceOf[Workflow.Effect[_]]
-    } else {
 
-      val invocationContext =
-        InvocationContext(
-          command.asInstanceOf[ScalaPbAny],
-          commandHandler.requestMessageDescriptor,
-          commandContext.metadata())
+    CommandResult(commandEffect)
+  }
+
+  /** INTERNAL API */
+  // "public" api against the impl/testkit
+  final def _internalHandleStep(
```

Comment:
The whole router is internal so we don't have to use `_internal` naming

Reply:
Sure. I'm changing it for workflows. The other routers also have those `_internal` methods. This come from the previous gRPC routers. There were some public APIs to be implemented.